### DISCUSSION TITLE
Variant for STM32H503CBT6, DAC not working, Help Needed.

### DISCUSSION THREAD

**Discussion by r-pv** [stm32duino/Arduino_Core_STM32#2268] [Variants] [answered]
Hi, so I'm trying to make a arduino variant for this MCU, it seems that was working until I tried the DAC, the board freezes when I use the DAC, while on STM32Cube IDE works properly, so don't know if there's a bug on this whole line (I tried the build-in RBT variant without success either) or an specific error on my variant. I'm using ST core 2.7.1

Schematic: https://github.com/r-pv/STM32H503CBT6/blob/main/STM32H503CBT6/STM32H503CBT6.pdf
Variant: https://github.com/r-pv/STM32H503CBT6/tree/main/H503CBT
STM32CubeIDE project:
[STM32H503_DAC.zip](https://github.com/stm32duino/Arduino_Core_STM32/files/14120129/STM32H503_DAC.zip)

So I can't see any error on my variant, can you spot any error?.

On arduino I'm just using a very simple sketch:

```
short c=0;
bool b=false;

void setup(){
	analogWriteResolution(12);
	pinMode(LED_BUILTIN,OUTPUT);
	digitalWrite(LED_BUILTIN,LOW);
	pinMode(PA4,OUTPUT);
	analogWrite(PA4,2048);
	delay(5000);
}

// the loop function runs over and over again forever
void loop(){
	digitalWrite(LED_BUILTIN,!b);
	b=!b;
	analogWrite(PA4,c);
	delay(1000);
	c=(c+128>4096)?0:c+128;
}
```
The Led just to see if the board is OK or froze, and the led is ON constantly, so freezes when the analogwrite is called.

The compiler didn't gave any error or warnings, On the variant the DAC is enabled by default but also used the hal_conf_extra.h and build_opt files just to be sure without differences.

Any help is appreciated.

**Comment by fpistm** [comment]
Hi @r-pv 
will try when got a Nucleo H503. I've tested with Nucleo H563and it works.

**Reply by r-pv** [reply]
thanks!

**Reply by fpistm** [reply]
Unfortunately, I didn't find a Nucelo H503. You should try to debug then see where it stuck.

**Reply by r-pv** [reply]
Thanks, I think this line was released on September, so it's pretty new.

I think the problem is related to the ADC/DAC clock settings, first time I tried on CubeIDE also got it froze due to ADC/DAC clock settings. After set this right the cube IDE generates the **SystemClock_Config** and **PeriphCommonClock_Config** functions, I saw on a build-in variant that this can be combined on just the **SystemClock_Config** function on the **generic_clock.c** file, but don't know if its the right way to do it or if the ADC/DAC clock settings are  being ignored/rewritten by the core.

This is my clock config on the variant:

```
WEAK void SystemClock_Config(void)
{
  RCC_OscInitTypeDef RCC_OscInitStruct = {};
  RCC_ClkInitTypeDef RCC_ClkInitStruct = {};
  RCC_PeriphCLKInitTypeDef PeriphClkInitStruct = {};

  /** Configure the main internal regulator output voltage
  */
  __HAL_PWR_VOLTAGESCALING_CONFIG(PWR_REGULATOR_VOLTAGE_SCALE0);

  while(!__HAL_PWR_GET_FLAG(PWR_FLAG_VOSRDY)) {}

  /** Initializes the RCC Oscillators according to the specified parameters
  * in the RCC_OscInitTypeDef structure.
  */
  RCC_OscInitStruct.OscillatorType = RCC_OSCILLATORTYPE_LSI|RCC_OSCILLATORTYPE_HSE;
  RCC_OscInitStruct.HSEState = RCC_HSE_ON;
  RCC_OscInitStruct.LSIState = RCC_LSI_ON;
  RCC_OscInitStruct.PLL.PLLState = RCC_PLL_ON;
  RCC_OscInitStruct.PLL.PLLSource = RCC_PLL1_SOURCE_HSE;
  RCC_OscInitStruct.PLL.PLLM = 3;
  RCC_OscInitStruct.PLL.PLLN = 125;
  RCC_OscInitStruct.PLL.PLLP = 2;
  RCC_OscInitStruct.PLL.PLLQ = 2;
  RCC_OscInitStruct.PLL.PLLR = 2;
  RCC_OscInitStruct.PLL.PLLRGE = RCC_PLL1_VCIRANGE_2;
  RCC_OscInitStruct.PLL.PLLVCOSEL = RCC_PLL1_VCORANGE_WIDE;
  RCC_OscInitStruct.PLL.PLLFRACN = 0;
  if (HAL_RCC_OscConfig(&RCC_OscInitStruct) != HAL_OK)
  {
    Error_Handler();
  }

  /** Initializes the CPU, AHB and APB buses clocks
  */
  RCC_ClkInitStruct.ClockType = RCC_CLOCKTYPE_HCLK|RCC_CLOCKTYPE_SYSCLK
                              |RCC_CLOCKTYPE_PCLK1|RCC_CLOCKTYPE_PCLK2
                              |RCC_CLOCKTYPE_PCLK3;
  RCC_ClkInitStruct.SYSCLKSource = RCC_SYSCLKSOURCE_PLLCLK;
  RCC_ClkInitStruct.AHBCLKDivider = RCC_SYSCLK_DIV1;
  RCC_ClkInitStruct.APB1CLKDivider = RCC_HCLK_DIV1;
  RCC_ClkInitStruct.APB2CLKDivider = RCC_HCLK_DIV1;
  RCC_ClkInitStruct.APB3CLKDivider = RCC_HCLK_DIV1;

  if (HAL_RCC_ClockConfig(&RCC_ClkInitStruct, FLASH_LATENCY_5) != HAL_OK)
  {
    Error_Handler();
  }

  /** Initializes the peripherals clock
  */
  PeriphClkInitStruct.PeriphClockSelection = RCC_PERIPHCLK_USB|RCC_PERIPHCLK_ADCDAC
                              |RCC_PERIPHCLK_SPI1;
  PeriphClkInitStruct.PLL2.PLL2Source = RCC_PLL2_SOURCE_HSE;
  PeriphClkInitStruct.PLL2.PLL2M = 1;
  PeriphClkInitStruct.PLL2.PLL2N = 12;
  PeriphClkInitStruct.PLL2.PLL2P = 1;
  PeriphClkInitStruct.PLL2.PLL2Q = 2;
  PeriphClkInitStruct.PLL2.PLL2R = 4;
  PeriphClkInitStruct.PLL2.PLL2RGE = RCC_PLL2_VCIRANGE_3;
  PeriphClkInitStruct.PLL2.PLL2VCOSEL = RCC_PLL2_VCORANGE_WIDE;
  PeriphClkInitStruct.PLL2.PLL2FRACN = 4096;
  PeriphClkInitStruct.PLL2.PLL2ClockOut = RCC_PLL2_DIVP|RCC_PLL2_DIVR;
  PeriphClkInitStruct.AdcDacClockSelection = RCC_ADCDACCLKSOURCE_PLL2R;
  PeriphClkInitStruct.Spi1ClockSelection = RCC_SPI1CLKSOURCE_PLL2P;
  PeriphClkInitStruct.UsbClockSelection = RCC_USBCLKSOURCE_PLL2Q;
  if (HAL_RCCEx_PeriphCLKConfig(&PeriphClkInitStruct) != HAL_OK)
  {
    Error_Handler();
  }

}
```

**Reply by fpistm** [reply]
it is the correct way. the clock config is not rewritten.

**Reply by r-pv** [reply, marked as answer]
Solved from your code on the thread:

https://github.com/stm32duino/Arduino_Core_STM32/issues/2508
https://github.com/fpistm/Arduino_Core_STM32/commit/170c4720df4722fa31fb5927ec624a1e6754746f

That was the issue, now the DAC works and not freezes anymore. 

✌ thanks.

**Reply by r-pv** [reply]
Sorry again, the mcu doesn't freeze anymore and the dac appears to work, but I don't get consistent steps between counts, it didn't matter if is set at 8bit or 12bit, it just doesn't appear to really work. I'm using this sketch:

```
const unsigned long ti=1000;
unsigned long t=ti;
short d=0;
bool b=false;




void setup(){
//	analogWriteResolution(12);
	pinMode(LED_BUILTIN,OUTPUT);
	digitalWrite(LED_BUILTIN,LOW);
	pinMode(PA4,OUTPUT);
//	analogWrite(PA4,0);

//	Serial.begin(115200);Serial.println();Serial.println();Serial.println();
}





void loop(){
	unsigned long _c=millis();
	if(_c>=t){
		d++;
		d=(d>255)?0:d;
//		Serial.println(d);

		analogWrite(PA4,d);

		t=_c+ti;
		b=!b;
		digitalWrite(LED_BUILTIN,b);}
}
```

with the multimeter on PA4 didn't show anything for several counts until above the 30count, then seems to work but stops after a few counts more, then start to move again and stops, and so on, if I reset the MCU sometimes it start to showing like: 1.5V, other times 1.2V others 0V. 

On STM32CubeIDE works as expected... the multimeter shows a floor of 12-16mV (I'm working on breadboard), until it reaches this floor start to adds ~800µV consistently, the sketch on STM32CubeIDE it's basically the same:

```
int main(void)
{
  /* USER CODE BEGIN 1 */

	uint16_t c=0;
	uint8_t b=0;

  /* USER CODE END 1 */

  /* MCU Configuration--------------------------------------------------------*/

  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
  HAL_Init();

  /* USER CODE BEGIN Init */

  /* USER CODE END Init */

  /* Configure the system clock */
  SystemClock_Config();

/* Configure the peripherals common clocks */
  PeriphCommonClock_Config();

  /* USER CODE BEGIN SysInit */

  /* USER CODE END SysInit */

  /* Initialize all configured peripherals */
  MX_GPIO_Init();
  MX_DAC1_Init();
  MX_ICACHE_Init();
  MX_SPI1_Init();
  MX_USB_PCD_Init();
  /* USER CODE BEGIN 2 */
  HAL_GPIO_WritePin(GPIOC,GPIO_PIN_13,GPIO_PIN_RESET);
  HAL_DAC_SetValue(&hdac1,DAC_CHANNEL_1,DAC_ALIGN_12B_R,0);
  /* USER CODE END 2 */

  /* Infinite loop */
  /* USER CODE BEGIN WHILE */
  while (1)
  {
    /* USER CODE END WHILE */
		if(b){
			b=0;
			HAL_GPIO_WritePin(GPIOC,GPIO_PIN_13,GPIO_PIN_RESET);
		}else{
			b=1;
			HAL_GPIO_WritePin(GPIOC,GPIO_PIN_13,GPIO_PIN_SET);}
		c++;
		c=(c>4095)?0:c;
		HAL_DAC_SetValue(&hdac1,DAC_CHANNEL_1,DAC_ALIGN_12B_R,c);
		HAL_Delay(1000);
    /* USER CODE BEGIN 3 */
  }
  /* USER CODE END 3 */
}
```

On Arduino IDE it just seems to act just randomly, so I need a bit of help again. Thanks.
